{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Imports"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "from Bio import SeqIO\n",
    "import matplotlib.pyplot as plt\n",
    "import matplotlib.cm as cm\n",
    "import seaborn as sns\n",
    "import pandas as pd\n",
    "from scipy.stats import wilcoxon, spearmanr\n",
    "from matplotlib.ticker import FuncFormatter, ScalarFormatter, LogLocator, LogFormatter\n",
    "import numpy as np\n",
    "from sklearn.metrics import f1_score\n",
    "import statsmodels.formula.api as smf\n",
    "from statsmodels.stats.multitest import multipletests"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Constants"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Constants\n",
    "SAMPLES = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20]\n",
    "ABUNDANCES = [1, 10, 20, 30, 40, 50, 60, 70, 80, 90, 100]\n",
    "METHODS = [\n",
    "    \"all\",\n",
    "\t\"centroid\",\n",
    "    \"ggrasp\",\n",
    "    \"vlq\",\n",
    "\t\"meshclust_0.95\",\n",
    "\t\"meshclust_0.99\",\n",
    "    \"gclust_0.95\",\n",
    "\t\"gclust_0.99\",\n",
    "\t\"gclust_0.999\",\n",
    "    \"vsearch_0.95\",\n",
    "    \"vsearch_0.99\",\n",
    "    \"vsearch_0.999\",\n",
    "    \"single-linkage_1\",\n",
    "    \"single-linkage_5\",\n",
    "    \"single-linkage_10\",\n",
    "    \"single-linkage_25\",\n",
    "    \"single-linkage_50\",\n",
    "    \"single-linkage_90\",\n",
    "    \"single-linkage_99\",\n",
    "\t\"complete-linkage_1\",\n",
    "    \"complete-linkage_5\",\n",
    "    \"complete-linkage_10\",\n",
    "    \"complete-linkage_25\",\n",
    "    \"complete-linkage_50\",\n",
    "    \"complete-linkage_90\",\n",
    "    \"complete-linkage_99\",\n",
    "\n",
    "]\n",
    "METHOD_LABELS = {\n",
    "    \"all\": \"All\",\n",
    "\t\"centroid\": \"Centroid\",\n",
    "    \"ggrasp\": \"GGRaSP\",\n",
    "    \"vlq\": \"VLQ\",\n",
    "\t\"meshclust_0.95\": \"MC-0.95\",\n",
    "\t\"meshclust_0.99\": \"MC-0.99\",\n",
    "    \"gclust_0.95\": \"GC-0.95\",\n",
    "\t\"gclust_0.99\": \"GC-0.99\",\n",
    "\t\"gclust_0.999\": \"GC-0.999\",\n",
    "    \"vsearch_0.95\": \"VS-0.95\",\n",
    "    \"vsearch_0.99\": \"VS-0.99\",\n",
    "    \"vsearch_0.999\": \"VS-0.999\",\n",
    "    \"single-linkage_1\": r\"SL-$P_{1}$\",\n",
    "    \"single-linkage_5\": r\"SL-$P_{5}$\",\n",
    "    \"single-linkage_10\": r\"SL-$P_{10}$\",\n",
    "    \"single-linkage_25\": r\"SL-$P_{25}$\",\n",
    "    \"single-linkage_50\": r\"SL-$P_{50}$\",\n",
    "    \"single-linkage_90\": r\"SL-$P_{90}$\",\n",
    "    \"single-linkage_99\": r\"SL-$P_{99}$\",\n",
    "\t\"complete-linkage_1\": r\"CL-$P_{1}$\",\n",
    "    \"complete-linkage_5\": r\"CL-$P_{5}$\",\n",
    "    \"complete-linkage_10\": r\"CL-$P_{10}$\",\n",
    "    \"complete-linkage_25\": r\"CL-$P_{25}$\",\n",
    "    \"complete-linkage_50\": r\"CL-$P_{50}$\",\n",
    "    \"complete-linkage_90\": r\"CL-$P_{90}$\",\n",
    "    \"complete-linkage_99\": r\"CL-$P_{99}$\",\n",
    "}\n",
    "EXPERIMENTS = [\"global\", \"country\", \"state\"]\n",
    "ALPHA = 0.05\n",
    "OUTPUT_DIR = \"figures/viruses\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Helper functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def create_mapping(experiment):\n",
    "    \"\"\"\n",
    "    Determine all lineages that are present in reference and simulation genomes\n",
    "    \"\"\"\n",
    "    idx2lineage = []\n",
    "    lineage2idx = {}\n",
    "    # Add lineages from reference\n",
    "    with open(f\"root/reference_sets/{experiment}/all.tsv\", \"r\") as f_in:\n",
    "        next(f_in) #skip header\n",
    "        for line in f_in:\n",
    "            line = line.strip().split(\"\\t\")\n",
    "            lineage = line[0]\n",
    "            if lineage not in lineage2idx:\n",
    "                lineage2idx[lineage] = len(idx2lineage)\n",
    "                idx2lineage.append(lineage)\n",
    "    print(len(lineage2idx), \"unique lineages found in reference\")\n",
    "    # Add lineages from simulations\n",
    "    with open(f\"root/samples/1/metadata.tsv\", \"r\") as f_in:\n",
    "        next(f_in)\n",
    "        for line in f_in:\n",
    "            line = line.strip().split(\"\\t\")\n",
    "            lineage = line[11]\n",
    "            if lineage not in lineage2idx:\n",
    "                lineage2idx[lineage] = len(idx2lineage)\n",
    "                idx2lineage.append(lineage)\n",
    "    print(len(lineage2idx), \"unique lineages found in total\")\n",
    "\n",
    "    return idx2lineage, lineage2idx\n",
    "\n",
    "def generate_groundtruth(idx2lineage, lineage2idx):\n",
    "    # Map sequence ids to lineages\n",
    "    id2lineage = {}\n",
    "    with open(f\"root/samples/1/metadata.tsv\", \"r\") as f_in:\n",
    "        next(f_in)\n",
    "        for line in f_in:\n",
    "            line = line.strip().split(\"\\t\")\n",
    "            id = line[0]\n",
    "            lineage = line[11]\n",
    "            id2lineage[id] = lineage\n",
    "\n",
    "    abundances = np.zeros((len(SAMPLES)*len(ABUNDANCES), len(idx2lineage)), dtype=np.float64)\n",
    "    for sample in SAMPLES:\n",
    "        # Map sequence ids to genome length\n",
    "        id2length = {}\n",
    "        for record in SeqIO.parse(f\"root/samples/{sample}/sequences.tsv\", \"fasta\"):\n",
    "            id2length[record.id.strip().split(\"|\")[0]] = len(record.seq)\n",
    "        for record in SeqIO.parse(\"root/B.1.1.7_sequence.fasta\", \"fasta\"):\n",
    "            id2length[record.id.strip().split(\"|\")[0]] = len(record.seq)\n",
    "\n",
    "        for i, abundance in enumerate(ABUNDANCES):\n",
    "            nucleotides = {id: 0 for id in id2length.keys()}\n",
    "            linecount = 0\n",
    "            with open(f\"root/samples/{sample}/wwsim_B.1.1.7_sequence_ab{abundance}_1.fastq\", \"r\") as f_in:\n",
    "                for line in f_in:\n",
    "                    if linecount % 4 == 0:\n",
    "                        id = line[1:].strip().split(\"|\")[0]\n",
    "                    elif linecount % 4 == 1:\n",
    "                        nucleotides[id] += len(line.strip())             \n",
    "                    linecount += 1\n",
    "            linecount = 0\n",
    "            with open(f\"root/samples/{sample}/wwsim_B.1.1.7_sequence_ab{abundance}_2.fastq\", \"r\") as f_in:\n",
    "                for line in f_in:\n",
    "                    if linecount % 4 == 0:\n",
    "                        id = line[1:].strip().split(\"|\")[0]\n",
    "                    elif linecount % 4 == 1:\n",
    "                        nucleotides[id] += len(line.strip())             \n",
    "                    linecount += 1\n",
    "\n",
    "            for id in nucleotides:\n",
    "                try:\n",
    "                    lineage = id2lineage[id]\n",
    "                except: # not included in sequences file!\n",
    "                    lineage = \"B.1.1.7\"\n",
    "                abundances[sample-1 + i*len(SAMPLES), lineage2idx[lineage]] += nucleotides[id] / id2length[id]\n",
    "            abundances[sample-1 + i*len(SAMPLES), :] = abundances[sample-1 + i*len(SAMPLES), :] / np.sum(abundances[sample-1 + i*len(SAMPLES), :])\n",
    "\n",
    "    return abundances\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Generate groundtruth"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": []
  }
 ],
 "metadata": {
  "language_info": {
   "name": "python"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
