{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Imports"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "from Bio import SeqIO\n",
    "import matplotlib.pyplot as plt\n",
    "import matplotlib.cm as cm\n",
    "import seaborn as sns\n",
    "import pandas as pd\n",
    "from scipy.stats import wilcoxon, spearmanr\n",
    "from matplotlib.ticker import FuncFormatter, ScalarFormatter, LogLocator, LogFormatter\n",
    "import numpy as np\n",
    "from sklearn.metrics import f1_score\n",
    "import statsmodels.formula.api as smf\n",
    "from statsmodels.stats.multitest import multipletests"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Constants"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Constants\n",
    "EXPERIMENTS = [\n",
    "    \"order_experiments\",\n",
    "    \"family_experiments\",\n",
    "    \"genus_experiments\"\n",
    "]\n",
    "SAMPLES = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]\n",
    "METHODS = [\n",
    "    \"all\",\n",
    "\t\"centroid\",\n",
    "    \"ggrasp\",\n",
    "    \"meshclust_0.95\",\n",
    "\t\"meshclust_0.97\",\n",
    "\t\"meshclust_0.99\",\n",
    "    \"gclust_0.95\",\n",
    "\t\"gclust_0.97\",\n",
    "\t\"gclust_0.99\",\n",
    "    \"single-linkage_0.95\",\n",
    "\t\"single-linkage_0.97\",\n",
    "\t\"single-linkage_0.99\",\n",
    "\t\"complete-linkage_0.95\",\n",
    "\t\"complete-linkage_0.97\",\n",
    "\t\"complete-linkage_0.99\",\n",
    "]\n",
    "METHOD_LABELS = {\n",
    "    \"all\": \"All\",\n",
    "    \"centroid\": \"Centroid\",\n",
    "    \"ggrasp\": \"GGRaSP\",\n",
    "    \"meshclust_0.95\": \"MC-0.95\",\n",
    "    \"meshclust_0.97\": \"MC-0.97\",\n",
    "    \"meshclust_0.99\": \"MC-0.99\",\n",
    "    \"gclust_0.95\": \"GC-0.95\",\n",
    "    \"gclust_0.97\": \"GC-0.97\",\n",
    "    \"gclust_0.99\": \"GC-0.99\",\n",
    "    \"single-linkage_0.95\": \"SL-0.95\",\n",
    "    \"single-linkage_0.97\": \"SL-0.97\",\n",
    "    \"single-linkage_0.99\": \"SL-0.99\",\n",
    "    \"complete-linkage_0.95\": \"CL-0.95\",\n",
    "    \"complete-linkage_0.97\": \"CL-0.97\",\n",
    "    \"complete-linkage_0.99\": \"CL-0.99\"\n",
    "}\n",
    "ALPHA = 0.05\n",
    "OUTPUT_DIR = \"figures/bacteria\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Fetch data and helper functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "data_per_experiment = {\n",
    "    \"order_experiments\": {},\n",
    "    \"family_experiments\": {},\n",
    "    \"genus_experiments\": {}\n",
    "}\n",
    "\n",
    "def map_ids(experiment):\n",
    "    \"\"\"\n",
    "    This function maps sequence IDs to accession numbers, species taxids, and lengths.\n",
    "    It also creates mappings for species to indices and vice versa.\n",
    "    \"\"\"\n",
    "    folder = f\"root/simulation_genomes/{experiment}\"\n",
    "\n",
    "    id2species = {}\n",
    "    species2idx = {}\n",
    "    idx2species = []\n",
    "    id2length = {}\n",
    "    id2accession = {}\n",
    "    accessions = set()\n",
    "\n",
    "    # Iterate through genome folders\n",
    "    for species in os.listdir(f\"{folder}\"):\n",
    "        for accession in os.listdir(f\"{folder}/{species}\"):\n",
    "            if accession.endswith(\".fna\"):\n",
    "                taxid = int(species)\n",
    "                length = 0\n",
    "                accessions.add(accession)\n",
    "                with open(f\"{folder}/{species}/{accession}\", \"r\") as f_in:\n",
    "                    for record in SeqIO.parse(f_in, \"fasta\"):\n",
    "                        length += len(record.seq)\n",
    "                        id2species[record.id] = taxid\n",
    "                        id2length[record.id] = len(record.seq)\n",
    "                        id2accession[record.id] = accession\n",
    "                        if taxid not in species2idx:\n",
    "                            species2idx[taxid] = len(idx2species)\n",
    "                            idx2species.append(taxid)\n",
    "                id2length[accession] = length\n",
    "                id2species[accession] = taxid\n",
    "    \n",
    "    return accessions, {\n",
    "        \"id2species\": id2species,\n",
    "        \"species2idx\": species2idx,\n",
    "        \"idx2species\": idx2species,\n",
    "        \"id2length\": id2length,\n",
    "        \"id2accession\": id2accession\n",
    "    }\n",
    "\n",
    "def index_species(experiment):\n",
    "    accession2species = {}\n",
    "    species2idx = {}\n",
    "    idx2species = []\n",
    "    folder = f\"root/reference_sets/{experiment}/all.tsv\"\n",
    "    with open(folder, \"r\") as f_in:\n",
    "        for line in f_in:\n",
    "            line = line.strip().split(\"\\t\")\n",
    "            taxid = int(line[1])\n",
    "            accession = line[2]\n",
    "            accession2species[accession] = taxid\n",
    "\n",
    "            if taxid not in species2idx:\n",
    "                species2idx[taxid] = len(idx2species)\n",
    "                idx2species.append(taxid)\n",
    "\n",
    "    return accession2species, species2idx, idx2species\n",
    "\n",
    "# This fetches all data and mappings for each experiment\n",
    "data_per_experiment = {}\n",
    "accessions_per_experiment = {}\n",
    "for experiment in EXPERIMENTS:\n",
    "    accessions_per_experiment[experiment], data_per_experiment[experiment] = map_ids(experiment)\n",
    "    a2s, s2i, i2s = index_species(experiment)\n",
    "    data_per_experiment[experiment][\"accession2species\"] = a2s\n",
    "    data_per_experiment[experiment][\"species2idx\"] = s2i\n",
    "    data_per_experiment[experiment][\"idx2species\"] = i2s"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Generate groundtruths"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Generate groundtruths (this takes ~9 minutes to run)\n",
    "groundtruths_genomes = {} #this is corrected for the length of genomes -> DUDes\n",
    "groundtruths_reads = {} #this only considers reads assigned to species -> Bracken & Centrifuge\n",
    "for experiment in EXPERIMENTS:\n",
    "    groundtruths_genomes[experiment] = np.zeros((len(SAMPLES), len(data_per_experiment[experiment][\"idx2species\"])), dtype=np.float64)\n",
    "    groundtruths_reads[experiment] = np.zeros((len(SAMPLES), len(data_per_experiment[experiment][\"idx2species\"])), dtype=np.float64)\n",
    "    cur_path = f\"root/samples/{experiment}/\"\n",
    "    for sample in SAMPLES:\n",
    "        line_count = 0\n",
    "        nucleotide_counts = {species: {} for species in data_per_experiment[experiment][\"idx2species\"]}\n",
    "        read_counts = {species: 0 for species in data_per_experiment[experiment][\"idx2species\"]}\n",
    "        total_reads = 0\n",
    "        with open(f\"{cur_path}/sample_{sample}/sample_1.fq\", \"r\") as f_in: # calculate number of reads and nucleotides per species\n",
    "            for line in f_in:\n",
    "                if line_count % 4 == 0:\n",
    "                    cur_id = line[1:].strip().split(\"-\")[0]\n",
    "                    cur_accession = data_per_experiment[experiment][\"id2accession\"][cur_id]\n",
    "                    read_counts[data_per_experiment[experiment][\"id2species\"][cur_id]] += 1\n",
    "                    total_reads += 1\n",
    "                elif line_count % 4 == 1:\n",
    "                    cur_read = line.strip()\n",
    "                    if cur_accession not in nucleotide_counts[data_per_experiment[experiment][\"id2species\"][cur_accession]]:\n",
    "                        nucleotide_counts[data_per_experiment[experiment][\"id2species\"][cur_id]][cur_accession] = 0\n",
    "                    nucleotide_counts[data_per_experiment[experiment][\"id2species\"][cur_id]][cur_accession] += len(cur_read)\n",
    "                line_count += 1\n",
    "\n",
    "        # Account for genome lengths\n",
    "        for species in nucleotide_counts:\n",
    "            species_idx = data_per_experiment[experiment][\"species2idx\"][species]\n",
    "            for accession in nucleotide_counts[species]:\n",
    "                groundtruths_genomes[experiment][sample-1, species_idx] += nucleotide_counts[species][accession] / data_per_experiment[experiment][\"id2length\"][accession]\n",
    "        # Only consider number of reads assigned\n",
    "        for species in read_counts:\n",
    "            species_idx = data_per_experiment[experiment][\"species2idx\"][species]\n",
    "            groundtruths_reads[experiment][sample-1, species_idx] = read_counts[species] / total_reads\n",
    "            groundtruths_genomes[experiment][sample-1, :] = groundtruths_genomes[experiment][sample-1, :] / np.sum(groundtruths_genomes[experiment][sample-1, :])\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Helper functions for fetching abundance estimates per profiler"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def read_bracken(experiment, method, data_per_experiment):\n",
    "    readcounts = np.zeros((len(SAMPLES), len(data_per_experiment[experiment][\"idx2species\"])), dtype=np.int64)\n",
    "    abundances = np.zeros((len(SAMPLES), len(data_per_experiment[experiment][\"idx2species\"])), dtype=np.float64)\n",
    "    for sample in SAMPLES:\n",
    "        total_readcount = 0\n",
    "        with open(f\"root/estimations/{experiment}/bracken/sample_${sample}/${method}.bracken\", \"r\") as f_in:\n",
    "            next(f_in) #skip header\n",
    "            for line in f_in:\n",
    "                line = line.strip().split(\"\\t\")\n",
    "                taxid = int(line[1])\n",
    "                taxid_idx = data_per_experiment[experiment][\"species2idx\"][taxid]\n",
    "                readcount = int(line[5])\n",
    "                readcounts[sample-1, taxid_idx] += readcount\n",
    "                total_readcount += readcount\n",
    "        for taxid_idx in range(len(data_per_experiment[experiment][\"idx2species\"])):\n",
    "            if readcounts[sample-1, taxid_idx] / total_readcount > 0.001:\n",
    "                abundances[sample-1, taxid_idx] = readcounts[sample-1, taxid_idx] / total_readcount\n",
    "        abundances[sample-1, :] = abundances[sample-1, :] / np.sum(abundances[sample-1, :])\n",
    "    return abundances\n",
    "\n",
    "def read_centrifuge(experiment, method, data_per_experiment):\n",
    "    readcounts = np.zeros((len(SAMPLES), len(data_per_experiment[experiment][\"idx2species\"])), dtype=np.int64)\n",
    "    abundances = np.zeros((len(SAMPLES), len(data_per_experiment[experiment][\"idx2species\"])), dtype=np.float64)\n",
    "    for sample in SAMPLES:\n",
    "        total_readcount = 0\n",
    "        with open(f\"root/estimations/{experiment}/centrifuge/sample_{sample}/{method}.report\", \"r\") as f_in:\n",
    "            next(f_in) #skip header\n",
    "            for line in f_in:\n",
    "                line = line.strip().split(\"\\t\")\n",
    "                taxid = int(line[1])\n",
    "                level = line[2]\n",
    "                if level == \"species\":\n",
    "                    taxid_idx = data_per_experiment[experiment][\"species2idx\"][taxid]\n",
    "                    readcount = int(line[5]) #line[5] is the number of reads uniquely assigned to this taxon\n",
    "                    readcounts[sample-1, taxid_idx] += readcount\n",
    "                    total_readcount += int(line[5])\n",
    "        for taxid_idx in range(len(data_per_experiment[experiment][\"idx2species\"])):\n",
    "            if readcounts[sample-1, taxid_idx] / total_readcount > 0.001:\n",
    "                abundances[sample-1, taxid_idx] = readcounts[sample-1, taxid_idx] / total_readcount\n",
    "        abundances[sample-1, :] = abundances[sample-1, :] / np.sum(abundances[sample-1, :])\n",
    "    return abundances\n",
    "\n",
    "def read_dudes(experiment, method, data_per_experiment):\n",
    "    abundances = np.zeros((len(SAMPLES), len(data_per_experiment[experiment][\"idx2species\"])), dtype=np.float64)\n",
    "    for sample in SAMPLES:\n",
    "        total_abundance = 0\n",
    "        with open(f\"root/estimations/{experiment}/dudes/sample_{sample}/{method}_dudes.out\", \"r\") as f_in:\n",
    "            for _ in range(6): #skip preamble\n",
    "                next(f_in)\n",
    "            for line in f_in:\n",
    "                line = line.strip().split(\"\\t\")\n",
    "                taxid = int(line[0])\n",
    "                level = line[1]\n",
    "                if level == \"species\":\n",
    "                    taxid_idx = data_per_experiment[experiment][\"species2idx\"][taxid]\n",
    "                    abundance = float(line[-1]) / 100 #scale to [0, 1]\n",
    "                    if abundance > 0.001:\n",
    "                        abundances[sample-1, taxid_idx] += abundance\n",
    "                        total_abundance += abundance\n",
    "        if total_abundance == 0:\n",
    "            print(experiment, method, sample)\n",
    "            abundances[sample-1, :] = np.nan\n",
    "        else:\n",
    "            abundances[sample-1, :] = abundances[sample-1, :] / np.sum(abundances[sample-1, :])\n",
    "    return abundances\n",
    "\n",
    "        "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Plotting functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "import matplotlib.cm as cm\n",
    "import numpy as np\n",
    "import seaborn as sns\n",
    "import pandas as pd\n",
    "from scipy.stats import wilcoxon, spearmanr\n",
    "import pandas as pd\n",
    "from matplotlib.ticker import FuncFormatter, ScalarFormatter, LogLocator, LogFormatter\n",
    "\n",
    "\n",
    "COLOR_PALETTE = {\n",
    "    \"1 - L1/2\": sns.color_palette(\"deep\")[0],\n",
    "    \"F1\": sns.color_palette(\"muted\")[1]\n",
    "}\n",
    "\n",
    "def plot_accuracy(experiment, metrics, results, pvalues_l1, pvalues_f1):\n",
    "    \"\"\"\n",
    "    This function plots the accuracy barplots for each profiler and experiment (Figure 3 in manuscript).\n",
    "    \"\"\"\n",
    "    for profiler in [\"bracken\", \"centrifuge\", \"dudes\"]:\n",
    "        df = pd.DataFrame(metrics[profiler])\n",
    "        g = sns.catplot(\n",
    "            data=df,\n",
    "            kind=\"bar\",\n",
    "            y=\"method\",\n",
    "            x=\"value\",\n",
    "            hue=\"metric\",\n",
    "            estimator=\"median\",\n",
    "            errorbar=(\"pi\", 80),\n",
    "            err_kws={\"linewidth\": 4},\n",
    "            width=0.9,\n",
    "            alpha=0.8,\n",
    "            order=[METHOD_LABELS[m] for m in METHODS[1:]],\n",
    "            palette=COLOR_PALETTE,\n",
    "        )\n",
    "        g.set_axis_labels(\"Score\", \"\", fontsize=16)\n",
    "        g.ax.axvline(x=np.median(results[profiler][\"all\"][\"L1\"]), ymin=0, ymax=1, color=COLOR_PALETTE[\"1 - L1/2\"], linestyle=\"--\", alpha=1, linewidth=2)\n",
    "        g.ax.axvline(x=np.median(results[profiler][\"all\"][\"F1\"]), ymin=0, ymax=1, color=COLOR_PALETTE[\"F1\"], linestyle=\"--\", alpha=1, linewidth=2)\n",
    "        g.ax.grid(True, which=\"major\", axis=\"x\", linestyle=\":\", alpha=0.5)\n",
    "\n",
    "        if profiler == \"bracken\":\n",
    "            title_str = f\"Bracken - {experiment.split('_')[0].capitalize()}\"\n",
    "            plt.title(f\"Bracken - {experiment.split('_')[0].capitalize()}\", fontsize=20)\n",
    "        elif profiler == \"centrifuge\":\n",
    "            title_str = f\"Centrifuge - {experiment.split('_')[0].capitalize()}\"\n",
    "            plt.title(f\"Centrifuge - {experiment.split('_')[0].capitalize()}\", fontsize=20)\n",
    "        else:\n",
    "            title_str = f\"DUDes - {experiment.split('_')[0].capitalize()}\"\n",
    "            plt.title(f\"DUDes - {experiment.split('_')[0].capitalize()}\", fontsize=20)\n",
    "\n",
    "        plt.yticks(fontsize=15)\n",
    "        plt.xticks([0.0, 0.2, 0.4, 0.6, 0.8, 1.0], fontsize=13)\n",
    "        plt.xlim([0.0, 1.05])\n",
    "        # To account for the fact that DUDes didn't provide results for the genus experiments with the centroid selection\n",
    "        if not (experiment == \"genus_experiments\" and profiler == \"dudes\"):\n",
    "            methods_labels = [METHOD_LABELS[m] for m in METHODS[1:]]\n",
    "        else:\n",
    "            methods_labels = [METHOD_LABELS[m] for m in METHODS[2:]]\n",
    "        \n",
    "        cur_pvalues_l1 = pvalues_l1[profiler]\n",
    "        cur_pvalues_f1 = pvalues_f1[profiler]\n",
    "        for ax in g.axes.flat:\n",
    "            for metric, container in zip([\"1 - L1/2\", \"F1\"], ax.containers):\n",
    "                for method_label, bar in zip(methods_labels, container.patches):\n",
    "                    group = df[(df[\"method\"] == method_label) & (df[\"metric\"] == metric)]\n",
    "                    if group.empty:\n",
    "                        continue\n",
    "\n",
    "                    x_90 = np.percentile(group[\"value\"], 90)\n",
    "                    # Choose significant p-values\n",
    "                    pval_dict = cur_pvalues_f1 if metric == \"F1\" else cur_pvalues_l1\n",
    "                    if pval_dict.get(method_label, 1.0) < ALPHA:\n",
    "                        y_center = bar.get_y() + bar.get_height() / 2\n",
    "                        ax.text(\n",
    "                            x_90 + 0.025,\n",
    "                            y_center,\n",
    "                            f\"\\u2217\",\n",
    "                            #f\"{method_label}\",\n",
    "                            ha=\"left\",\n",
    "                            va=\"center\",\n",
    "                            fontsize=10,\n",
    "                            fontweight=\"bold\"\n",
    "                        )\n",
    "\n",
    "        # Fix legend\n",
    "        g.ax.legend(title=\"\", fontsize=15, ncol=2)\n",
    "\n",
    "        g.figure.set_size_inches(8, 6)\n",
    "        g.figure.tight_layout()\n",
    "        g.savefig(f\"{OUTPUT_DIR}/{experiment}/{profiler}_barplot.svg\", dpi=500, bbox_inches=\"tight\")\n",
    "        plt.close()\n",
    "\n",
    "def plot_genomes(experiment, results):\n",
    "    for profiler in [\"bracken\", \"centrifuge\", \"dudes\"]:\n",
    "        median_outcomes_l1 = np.array([np.median(results[profiler][method][\"L1\"]) for method in METHODS], dtype=np.float64)\n",
    "        median_outcomes_f1 = np.array([np.median(results[profiler][method][\"F1\"]) for method in METHODS], dtype=np.float64)\n",
    "        genome_counts = np.array([results[profiler][method][\"genomes\"] for method in METHODS], dtype=np.int64)\n",
    "\n",
    "        rho_l1, _ = spearmanr(genome_counts, median_outcomes_l1, nan_policy=\"omit\")\n",
    "        rho_f1, _ = spearmanr(genome_counts, median_outcomes_f1, nan_policy=\"omit\")\n",
    "\n",
    "        fig = plt.figure(figsize=(8,6))\n",
    "        ax = sns.scatterplot(x=genome_counts, y=median_outcomes_l1, s=100, color=sns.color_palette(\"deep\")[0], label=\"1 - L1/2\", edgecolor=\"black\", alpha=0.5)\n",
    "        sns.scatterplot(x=genome_counts, y=median_outcomes_f1, s=100, color=sns.color_palette(\"muted\")[1], label=\"F1\", edgecolor=\"black\", alpha=0.5)\n",
    "        \n",
    "        plt.xlabel(\"Number of genomes\", fontsize=16)\n",
    "        plt.ylabel(\"Score\", fontsize=16)\n",
    "        ax.grid(True, which=\"major\", axis=\"y\", linestyle=\":\", alpha=0.5)\n",
    "        plt.title(rf\"$\\rho(1-L1/2)={rho_l1:.3f}$, $\\rho(F1)={rho_f1:.3f}$\", fontsize=16)\n",
    "        plt.ylim([0, 1.05])\n",
    "        plt.xlim([0, max(genome_counts)+300])\n",
    "        plt.xticks(fontsize=13)\n",
    "        plt.yticks([0.0, 0.2, 0.4, 0.6, 0.8, 1.0], fontsize=13)\n",
    "        ax.spines[\"top\"].set_visible(False)\n",
    "        ax.spines[\"right\"].set_visible(False)\n",
    "        ax.legend(title=\"\", fontsize=15, ncol=2)\n",
    "        plt.tight_layout()\n",
    "        fig.savefig(f\"{OUTPUT_DIR}/{experiment}/{profiler}_l1_f1_genomes.svg\", dpi=500, bbox_inches=\"tight\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Determine L1 and F1 for all methods and plot\n",
    "**NOTE**: In this script we hardcode the experiments. This should be changed accordingly!!!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "experiment = \"genus_experiments\"  # Change this to run for different experiments!!\n",
    "\n",
    "results = {\n",
    "    \"bracken\": {},\n",
    "    \"centrifuge\": {},\n",
    "    \"dudes\": {}\n",
    "}\n",
    "metrics = {\n",
    "    \"bracken\": [],\n",
    "    \"centrifuge\": [],\n",
    "    \"dudes\": []\n",
    "}\n",
    "for method in METHODS:\n",
    "    bracken = read_bracken(experiment, method, data_per_experiment)\n",
    "    centrifuge = read_centrifuge(experiment, method, data_per_experiment)\n",
    "    dudes = read_dudes(experiment, method, data_per_experiment)\n",
    "\n",
    "    num_genomes = 0\n",
    "    with open(f\"root/reference_sets/{experiment}/{method}.tsv\", \"r\") as f_in:\n",
    "        for line in f_in:\n",
    "            num_genomes += 1\n",
    "\n",
    "    results[\"bracken\"][method] = {\n",
    "        \"L1\": [1.0 - np.sum( np.abs(bracken[sample, :] - groundtruths_reads[experiment][sample, :]) ) / 2.0 for sample in range(len(SAMPLES))],\n",
    "        \"F1\": [f1_score(groundtruths_reads[experiment][sample, :] > 0, bracken[sample, :] > 0, zero_division=np.nan) for sample in range(len(SAMPLES))],\n",
    "        \"genomes\": num_genomes,\n",
    "    }\n",
    "    results[\"centrifuge\"][method] = {\n",
    "        \"L1\": [1.0 - np.sum( np.abs(centrifuge[sample, :] - groundtruths_reads[experiment][sample, :]) ) / 2.0 for sample in range(len(SAMPLES))],\n",
    "        \"F1\": [f1_score(groundtruths_reads[experiment][sample, :] > 0, centrifuge[sample, :] > 0, zero_division=np.nan) for sample in range(len(SAMPLES))],\n",
    "        \"genomes\": num_genomes,\n",
    "    }\n",
    "    results[\"dudes\"][method] = {\n",
    "        \"L1\": [1.0 - np.sum( np.abs(dudes[sample, :] - groundtruths_genomes[experiment][sample, :]) ) / 2.0 for sample in range(len(SAMPLES))],\n",
    "        \"F1\": [f1_score(groundtruths_genomes[experiment][sample, :] > 0, dudes[sample, :] > 0, zero_division=np.nan) for sample in range(len(SAMPLES))],\n",
    "        \"genomes\": num_genomes,\n",
    "    }\n",
    "\n",
    "    for sample in range(len(SAMPLES)):\n",
    "        # Bracken\n",
    "        metrics[\"bracken\"].append(\n",
    "            {\n",
    "                \"method\": METHOD_LABELS[method],\n",
    "                \"metric\": \"1 - L1/2\",\n",
    "                \"value\": 1.0 - np.sum( np.abs(bracken[sample, :] - groundtruths_reads[experiment][sample, :]) ) / 2.0,\n",
    "                \"genomes\": num_genomes\n",
    "            }\n",
    "        )\n",
    "        metrics[\"bracken\"].append(\n",
    "            {\n",
    "                \"method\": METHOD_LABELS[method],\n",
    "                \"metric\": \"F1\",\n",
    "                \"value\": f1_score(groundtruths_reads[experiment][sample, :] > 0, bracken[sample, :] > 0, zero_division=np.nan),\n",
    "                \"genomes\": num_genomes\n",
    "            }\n",
    "        )\n",
    "        # Centrifuge\n",
    "        metrics[\"centrifuge\"].append(\n",
    "            {\n",
    "                \"method\": METHOD_LABELS[method],\n",
    "                \"metric\": \"1 - L1/2\",\n",
    "                \"value\": 1.0 - np.sum( np.abs(centrifuge[sample, :] - groundtruths_reads[experiment][sample, :]) ) / 2.0,\n",
    "                \"genomes\": num_genomes\n",
    "            }\n",
    "        )\n",
    "        metrics[\"centrifuge\"].append(\n",
    "            {\n",
    "                \"method\": METHOD_LABELS[method],\n",
    "                \"metric\": \"F1\",\n",
    "                \"value\": f1_score(groundtruths_reads[experiment][sample, :] > 0, centrifuge[sample, :] > 0, zero_division=np.nan),\n",
    "                \"genomes\": num_genomes\n",
    "            }\n",
    "        )\n",
    "        # DUDes\n",
    "        metrics[\"dudes\"].append(\n",
    "            {\n",
    "                \"method\": METHOD_LABELS[method],\n",
    "                \"metric\": \"1 - L1/2\",\n",
    "                \"value\": 1.0 - np.sum( np.abs(dudes[sample, :] - groundtruths_genomes[experiment][sample, :]) ) / 2.0,\n",
    "                \"genomes\": num_genomes\n",
    "            }\n",
    "        )\n",
    "        if method == \"centroid\" and experiment == \"genus_experiments\":\n",
    "            metrics[\"dudes\"].append(\n",
    "                {\n",
    "                    \"method\": METHOD_LABELS[method],\n",
    "                    \"metric\": \"F1\",\n",
    "                    \"value\": np.nan,\n",
    "                    \"genomes\": num_genomes\n",
    "                }\n",
    "            )\n",
    "        else:\n",
    "            metrics[\"dudes\"].append(\n",
    "                {\n",
    "                    \"method\": METHOD_LABELS[method],\n",
    "                    \"metric\": \"F1\",\n",
    "                    \"value\": f1_score(groundtruths_genomes[experiment][sample, :] > 0, dudes[sample, :] > 0, zero_division=np.nan),\n",
    "                    \"genomes\": num_genomes\n",
    "                }\n",
    "            )\n",
    "# Determine pvalues and significance\n",
    "pvalues_l1 = {}\n",
    "pvalues_corrected_l1 = {}\n",
    "pvalues_f1 = {}\n",
    "pvalues_corrected_f1 = {}\n",
    "for profiler in [\"bracken\", \"centrifuge\", \"dudes\"]:\n",
    "    cur_results = results[profiler]\n",
    "    cur_metrics = metrics[profiler]\n",
    "\n",
    "    cur_pvalues_l1 = np.ones(len(METHODS)-1, dtype=np.float64)\n",
    "    cur_pvalues_f1 = np.ones(len(METHODS)-1, dtype=np.float64)\n",
    "    for i, method in enumerate(METHODS[1:]):\n",
    "        if method == \"centroid\" and experiment == \"genus_experiments\" and profiler == \"dudes\":\n",
    "            continue\n",
    "        else:\n",
    "            # L1\n",
    "            try:\n",
    "                _, pval = wilcoxon(\n",
    "                    cur_results[method][\"L1\"],\n",
    "                    cur_results[\"all\"][\"L1\"],\n",
    "                    alternative=\"greater\"\n",
    "                )\n",
    "            except:\n",
    "                pval = 1.0\n",
    "            cur_pvalues_l1[i] = pval\n",
    "            #F1\n",
    "            try:\n",
    "                _, pval = wilcoxon(\n",
    "                    cur_results[method][\"F1\"],\n",
    "                    cur_results[\"all\"][\"F1\"],\n",
    "                    alternative=\"greater\"\n",
    "                )\n",
    "            except:\n",
    "                pval = 1.0\n",
    "            cur_pvalues_f1[i] = pval\n",
    "    # Multiple testing correction (Benjamini-Hochberg)\n",
    "    methods_for_pvalues = [METHOD_LABELS[m] for m in METHODS[1:]]\n",
    "    pvalues_l1[profiler] = cur_pvalues_l1\n",
    "    pvalues_corrected_l1[profiler] = dict(zip(methods_for_pvalues, multipletests(cur_pvalues_l1, method=\"fdr_bh\")[1]))\n",
    "    pvalues_f1[profiler] = cur_pvalues_f1\n",
    "    pvalues_corrected_f1[profiler] = dict(zip(methods_for_pvalues, multipletests(cur_pvalues_f1, method=\"fdr_bh\")[1]))\n",
    "\n",
    "\n",
    "plot_accuracy(experiment, metrics, results, pvalues_corrected_l1, pvalues_corrected_f1)\n",
    "plot_genomes(experiment, results)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": []
  }
 ],
 "metadata": {
  "language_info": {
   "name": "python"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
